### PR TITLE
gh-106844: Fix null-bytes handling in `LCMapStringEx` in `_winapi`

### PR DESCRIPTION
**Pull request by sobolevn**
I've changed the clinic definition + backported changes made by @zooba in https://github.com/python/cpython/commit/927b5afee73218500a8fa80df86216cfdc24ef5a#diff-42415407f8d0ef2d42e29d13d979f633e3543770e62c3871e1101ad532d336a8

* Issue: gh-106844

### CODE REVIEW

File: Modules/_winapi.c [hunk 1]
```diff
@@ -1537,42 +1537,66 @@ _winapi_PeekNamedPipe_impl(PyObject *module, HANDLE handle, int size)
 /*[clinic input]
 _winapi.LCMapStringEx
 
-    locale: LPCWSTR
+    locale: unicode
```
Comment (serhiy-storchaka) [inline Modules/_winapi.c]: Why change it for `locale`?

File: Modules/_winapi.c [hunk 2]
```diff
@@ -1537,42 +1537,66 @@ _winapi_PeekNamedPipe_impl(PyObject *module, HANDLE handle, int size)
...
+    int srcLen;
+    wchar_t *src_ = PyUnicode_AsWideCharString(src, &srcLenAsSsize);
+    if (!src_) {
+        PyMem_Free(locale_);
+        return NULL;
+    }
+    srcLen = (int)srcLenAsSsize;
+    if (srcLen != srcLenAsSsize) {
+        srcLen = -1;
+    }
```
Comment (serhiy-storchaka) [inline Modules/_winapi.c]: How would it work for large (more than `2**32` characters) string containing NUL?

File: Modules/_winapi.c (hunk 1 above)
Comment (sobolevn) [reply Modules/_winapi.c]: I just backported the solution from https://github.com/python/cpython/commit/927b5afee73218500a8fa80df86216cfdc24ef5a#diff-42415407f8d0ef2d42e29d13d979f633e3543770e62c3871e1101ad532d336a8R1518

So, 3.11 and 3.12 would match.

Comment (serhiy-storchaka) [reply Modules/_winapi.c]: The code would be simpler if keep the LPCWSTR converter here.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please address my comments. If the code in 3.11 has the same flaws, it should be fixed too.

**Comment by zooba** [comment]
I don't think the fix is as simple as bringing forward this fix - the underlying code should be identical in current versions, it was just earlier versions that needed a different fix because of existing bugs in the Unicode converters.

I'll take a fresh look and see what ought to happen here. Chances are, updating the `LPCWSTR` clinic definition makes the most sense.

File: Modules/_winapi.c (hunk 2 above)
Comment (zooba) [reply Modules/_winapi.c]: It doesn't, but that's the caller's responsibility.

**Comment by zooba** [comment]
Looks like the changes to make argument clinic support `Py_UNICODE_converter` actually work when you opt to allow embedded nulls are... complex. I can't figure it out, and pretty sure if I did it'd have wide-ranging effects that we don't want to backport.

Leaving the `locale` argument as `LPCWSTR` should be fine - we only need to handle embedded nulls in the `src` argument.

File: Modules/_winapi.c (hunk 1 above)
Comment (zooba) [reply Modules/_winapi.c]: I agree. We don't need to treat this argument any differently - it's only the `src` argument that requires special handling.

File: Modules/_winapi.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_winapi.c]: `LCMapStringEx()` will simply return the result for the truncated input. It is bad. How is the caller expected to detect this problem if it only occurs in special circumferences?

What does `LCMapStringEx()` return for large input not containing NUL? Most likely it fails. We should either emulate this response (raise an OSError with the same Windows error code) or simply raise an OverflowError.

**Comment by serhiy-storchaka** [comment]
Opened #107832 for my propositions.

**Comment by serhiy-storchaka** [comment]
#107832 fixed also other issues.

**Pull request closed without merging**